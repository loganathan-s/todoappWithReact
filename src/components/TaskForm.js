import React from 'react';
import PropTypes from 'prop-types';

class TaskForm extends React.Component{
	

	createTask(event){
		event.preventDefault();
		const task = {task: this.task.value, status: "New"}
		this.props.addTask(task);
		this.taskForm.reset();
	}	

	render(){
		return(
			 <form ref={(input) => this.taskForm = input} className="taskForm"  onSubmit={this.createTask.bind(this)}>
			  <input ref={(input) => this.task = input } type="text" placeholder="Task" />
			  <button type="submit" className="createTask">Add Task</button> 
			 </form>
			)
	}

}


TaskForm.PropTypes = {
  addTask: PropTypes.func.isRequired,
};

export default TaskForm;

