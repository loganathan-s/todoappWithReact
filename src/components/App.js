import React, { Component } from 'react';
import Header from './Header';
import Footer from './Footer';
import TaskForm from './TaskForm';
import Task from './Task';
import logo from '../images/logo.svg';
import {Tasks} from '../simpleTasks';

import '../css/App.css';

class App extends Component {
  constructor(){
    super();
    this.loadSamples = this.loadSamples.bind(this); 
    this.addTask = this.addTask.bind(this); 
    this.removeTask = this.removeTask.bind(this); 
    this.updateTask = this.updateTask.bind(this); 
    this.markComplete = this.markComplete.bind(this); 
    this.setCurrentTaskEditable = this.setCurrentTaskEditable.bind(this); 
    this.state = {
      tasks: {},
      editing: false,
      currentTask: {}
    };
  }

  addTask(task){
    const tasks = {...this.state.tasks};
    const taskKey = Object.keys(tasks).length + 1;
    tasks[`task${taskKey}`] = task;
    this.setState({ tasks })
  }

  removeTask(key){
     const tasks = {...this.state.tasks};
     delete tasks[key];
     this.setState({ tasks }) 
  }

  updateTask(key, updatedTask){
     const tasks = {...this.state.tasks};
     const task = tasks[key]
     task.task = updatedTask
     task.edit = false
     tasks[key] = task
     this.setState({ tasks }) 
  }

  markComplete(key){
     const tasks = {...this.state.tasks};
     let task = tasks[key]
     task.status = (task.status === "Done" ? "InProgress" : "Done")
     tasks[key] = task
     this.setState({ tasks }) 
  }

  setCurrentTaskEditable(key){
    const tasks = {...this.state.tasks};
    const currentTask = tasks[key]
    currentTask["edit"] = !currentTask["edit"];
    tasks[key] = currentTask;
    this.setState({ tasks });
  }

  loadSamples(){
      this.setState({
        tasks: Tasks
      })
  }

  //Callback which runs before rendering the APP component
  componentWillMount(){ 
    this.loadSamples();
  }

    
  render() {
    return (
      <div className="App">
         <Header />
          <div className="content">
            <TaskForm addTask={this.addTask} /> 
            <ul className="allTasks">
              { Object.
                keys(this.state.tasks).map(
                  key => <Task key={key} index={key} task={this.state.tasks[key]} removefromTask={this.removeTask} markasDone={this.markComplete}  setCurrentTaskEditable={this.setCurrentTaskEditable} updateTask={this.updateTask}/>)
            }
            </ul>
          </div>
        <Footer/>
      </div>
    );
  }
}

export default App;
