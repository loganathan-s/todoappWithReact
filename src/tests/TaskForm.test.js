import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import App from '../components/App';
import TaskForm from '../components/TaskForm';

describe(TaskForm, () => {
  
  const addTask = jest.fn();
  const createTask = jest.fn();
  const component = mount(<TaskForm addTask={addTask} onSubmit={createTask}/>);
  const appComponent = shallow(<App/>);

  it('renders and matches the snapshot', () => {
    	const component = renderer.create(<TaskForm addTask={addTask} />);
    	const tree = component.toJSON();
    	expect(tree).toMatchSnapshot();
  });

  it('should display the form', () => {
  	expect(component.find(".taskForm")).toHaveLength(1);
  });

  it('should create new task', () => {
    const totalTasks = Object.keys(appComponent.state('tasks')).length;
  	expect(component.find(".taskForm")).toHaveLength(1);
    const taskInput = component.find('input');
    taskInput.value = "New Task";
    expect(taskInput.value).toBe('New Task')
    component.find("form.taskForm").simulate('submit', { preventDefault(){} });
    expect(addTask).toBeCalled();
    appComponent.instance().addTask({task: "New Task", status: "New"})
    expect(Object.keys(appComponent.state('tasks')).length).toEqual(totalTasks + 1);
  });

});