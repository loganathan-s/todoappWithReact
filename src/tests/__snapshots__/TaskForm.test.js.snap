// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class TaskForm extends _react2.default.Component {


	createTask(event) {
		event.preventDefault();
		const task = { task: this.task.value, status: "New" };
		this.props.addTask(task);
		this.taskForm.reset();
	}

	render() {
		return (
			_react2.default.createElement('form', { ref: input => this.taskForm = input, className: 'taskForm', onSubmit: this.createTask.bind(this), __source: { fileName: _jsxFileName, lineNumber: 16 }, __self: this },
				_react2.default.createElement('input', { ref: input => this.task = input, type: 'text', placeholder: 'Task', __source: { fileName: _jsxFileName, lineNumber: 17 }, __self: this }),
				_react2.default.createElement('button', { type: 'submit', className: 'createTask', __source: { fileName: _jsxFileName, lineNumber: 18 }, __self: this }, 'Add Task')));


	}} renders and matches the snapshot 1`] = `
<form
  className="taskForm"
  onSubmit={[Function]}
>
  <input
    placeholder="Task"
    type="text"
  />
  <button
    className="createTask"
    type="submit"
  >
    Add Task
  </button>
</form>
`;
